packages/framework/src/client/resource-request.js

```diff
@@ -0,0 +1,111 @@
+import debugFactory from 'debug';
+import { isDateEarlier } from '../utils/dates';
+import { SECOND } from '../utils/constants';
+
+const DEFAULT_TIMEOUT = 30 * SECOND;
+
+export const STATUS = {
+	failed: 'Failed',
+	inFlight: 'In Flight',
+	complete: 'Complete',
+	overdue: 'Overdue',
+	scheduled: 'Scheduled',
+	timedOut: 'Timed Out',
+	unnecessary: 'Unnecessary',
+};
+
+export default class ResourceRequest {
+	constructor( resourceName, requirement, resourceState, now ) {
+		this.debug = debugFactory( 'fresh-data:request(' + resourceName + ')' );
+		this.resourceName = resourceName;
+		this.time = calculateRequestTime( requirement, resourceState, now );
+		this.timeout = requirement.timeout || DEFAULT_TIMEOUT;
+		this.promise = null;
+		this.timeRequested = null;
+		this.timeCompleted = null;
+		this.error = null;
+
+		if ( this.time ) {
+			const seconds = ( this.time.getTime() - now.getTime() ) / 1000.0;
+			this.debug( `New request for "${ this.resourceName }" to be fetched in ${ seconds }s` );
+		}
+	}
+
+	addRequirement( requirement, resourceState, now ) {
+		const status = this.getStatus( now );
+		if ( STATUS.scheduled === status || STATUS.overdue === status ) {
+			const requestTime = calculateRequestTime( requirement, resourceState, now );
+			if ( isDateEarlier( this.time, requestTime ) ) {
+				this.time = requestTime;
+				const seconds = ( this.time.getTime() - now.getTime() ) / 1000.0;
+				this.debug( `Rescheduling request for "${ this.resourceName }" to fetched in ${ seconds }s` );
+			}
+			this.timeout = Math.min( this.timeout, requirement.timeout || DEFAULT_TIMEOUT );
+			return true;
+		}
+		this.debug( `Cannot add requirement to request with "${ this.getStatus( now ) }" status` );
```

Comment:
Why not? If the request was in `STATUS.unnecessary` and now a new requirement means that the request will need to be scheduled, it should work, right? Or will a request that's `unnecessary` be immediately discarded and a new one would need to be created?

Reply:
It's a fair point that the unnecessary status is a bit confusing. I wonder if I can remove it altogether.

Reply:
I think for now, it's okay to leave this as-is. The scheduler shouldn't be creating unnecessary requests in the first place, so this is definitely in "should not happen" territory in practice.